miniapps/blocksolvers/div_free_solver.hpp

```diff
@@ -0,0 +1,252 @@
+#include "mfem.hpp"
+#include <memory>
+
+using namespace std;
+using namespace mfem;
+
+/// Parameters for iterative solver
+struct IterSolveParameters
+{
+   int print_level = 0;
+   int max_iter = 500;
+   double abs_tol = 1e-12;
+   double rel_tol = 1e-9;
+};
+
+/// Parameters for the divergence free solver
+struct DFSParameters : IterSolveParameters
+{
+   bool verbose = false;
+   bool B_has_nullity_one = false;   // whether B has a 1D nullspace
+   bool coupled_solve = false;       // whether to solve all unknowns together
+   IterSolveParameters BBT_solve_param;
+};
+
+/// Data for the divergence free solver
+struct DFSData
+{
+   Array<OperatorPtr> agg_hdivdof;    // agglomerates to H(div) dofs table
+   Array<OperatorPtr> agg_l2dof;      // agglomerates to L2 dofs table
+   Array<OperatorPtr> P_hdiv;         // Interpolation matrix for H(div) space
+   Array<OperatorPtr> P_l2;           // Interpolation matrix for L2 space
+   Array<OperatorPtr> P_hcurl;        // Interpolation for kernel space of div
+   Array<OperatorPtr> Q_l2;           // Q_l2[l] = (W_{l+1})^{-1} P_l2[l]^T W_l
+   Array<int> coarsest_ess_hdivdofs;  // coarsest level essential H(div) dofs
+   Array<OperatorPtr> C;              // discrete curl: ND -> RT
+   DFSParameters param;
+};
+
+/// For collecting data needed for the divergence free solver
+class DFSDataCollector
+{
+   RT_FECollection hdiv_fec_;
+   L2_FECollection l2_fec_;
+   ND_FECollection hcurl_fec_;
+   L2_FECollection l2_0_fec_;
+
+   unique_ptr<ParFiniteElementSpace> coarse_hdiv_fes_;
+   unique_ptr<ParFiniteElementSpace> coarse_l2_fes_;
+   unique_ptr<ParFiniteElementSpace> coarse_hcurl_fes_;
+   unique_ptr<ParFiniteElementSpace> l2_0_fes_;
+
+   Array<SparseMatrix> el_l2dof_;
+   const Array<int>& ess_bdr_attr_;
+   Array<int> all_bdr_attr_;
+
+   int level_;
+   int order_;
+   DFSData data_;
+
+   void MakeDofRelationTables(int level);
+   void DataFinalize();
+public:
+   DFSDataCollector(int order, int num_refine, ParMesh *mesh,
+                    const Array<int>& ess_attr, const DFSParameters& param);
+
+   void CollectData();
+
+   const DFSData& GetData() const { return data_; }
+
+   unique_ptr<ParFiniteElementSpace> hdiv_fes_;
+   unique_ptr<ParFiniteElementSpace> l2_fes_;
+   unique_ptr<ParFiniteElementSpace> hcurl_fes_;
+};
+
+/// Abstract solver class for Darcy's flow
+class DarcySolver : public Solver
+{
+protected:
+   Array<int> offsets_;
+public:
+   DarcySolver(int size0, int size1) : Solver(size0 + size1), offsets_(3)
+   { offsets_[0] = 0; offsets_[1] = size0; offsets_[2] = height; }
+   virtual int GetNumIterations() const = 0;
+};
+
+/// Solver for B * B^T
```

Comment:
More documentation (ie, specify that you explicitly assemble the matrix product and use BoomerAMG).